modules/message-lane/src/outbound_lane.rs

```diff
@@ -0,0 +1,269 @@
+
+		Some((prev_latest_received_nonce + 1, latest_received_nonce))
+	}
+
+	/// Confirm message processing.
+	///
+	/// Returns `None` if confirmation is wrong/duplicate.
+	/// Returns `Some` with inclusive ranges of message nonces that have been processed.
+	pub fn confirm_processing(&mut self, latest_processed_nonce: MessageNonce) -> Option<(MessageNonce, MessageNonce)> {
+		let mut data = self.storage.data();
+		// wait for recieval confirmation first
+		if latest_processed_nonce <= data.latest_processed_nonce || latest_processed_nonce > data.latest_received_nonce
+		{
+			return None;
+		}
+
+		let prev_latest_processed_nonce = data.latest_processed_nonce;
+		data.latest_processed_nonce = latest_processed_nonce;
+		self.storage.set_data(data);
+
+		Some((prev_latest_processed_nonce + 1, latest_processed_nonce))
+	}
+
+	/// Prune at most `max_messages_to_prune` already received messages.
+	pub fn prune_messages(&mut self, max_messages_to_prune: MessageNonce) -> MessageNonce {
+		let mut pruned_messages = 0;
+		let mut anything_changed = false;
+		let mut data = self.storage.data();
+		while pruned_messages < max_messages_to_prune && data.oldest_unpruned_nonce <= data.latest_received_nonce {
+			self.storage.remove_message(&data.oldest_unpruned_nonce);
+
+			anything_changed = true;
+			pruned_messages += 1;
+			data.oldest_unpruned_nonce += 1;
+		}
+
+		if anything_changed {
+			self.storage.set_data(data);
+		}
+
+		pruned_messages
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::{
+		mock::{run_test, TestRuntime, REGULAR_PAYLOAD, TEST_LANE_ID},
+		outbound_lane,
+	};
+
+	#[test]
+	fn send_message_works() {
+		run_test(|| {
+			let mut lane = outbound_lane::<TestRuntime, _>(TEST_LANE_ID);
+			assert_eq!(lane.storage.data().latest_generated_nonce, 0);
+			assert_eq!(lane.send_message(REGULAR_PAYLOAD), 1);
+			assert!(lane.storage.message(&1).is_some());
+			assert_eq!(lane.storage.data().latest_generated_nonce, 1);
+		});
+	}
+
+	#[test]
+	fn confirm_receival_works() {
+		run_test(|| {
+			let mut lane = outbound_lane::<TestRuntime, _>(TEST_LANE_ID);
+			assert_eq!(lane.send_message(REGULAR_PAYLOAD), 1);
+			assert_eq!(lane.send_message(REGULAR_PAYLOAD), 2);
+			assert_eq!(lane.send_message(REGULAR_PAYLOAD), 3);
+			assert_eq!(lane.storage.data().latest_generated_nonce, 3);
+			assert_eq!(lane.storage.data().latest_received_nonce, 0);
+			assert_eq!(lane.confirm_receival(3), Some((1, 3)));
+			assert_eq!(lane.storage.data().latest_generated_nonce, 3);
+			assert_eq!(lane.storage.data().latest_received_nonce, 3);
+		});
+	}
+
+	#[test]
+	fn confirm_receival_rejects_nonce_lesser_than_latest_received() {
+		run_test(|| {
+			let mut lane = outbound_lane::<TestRuntime, _>(TEST_LANE_ID);
+			lane.send_message(REGULAR_PAYLOAD);
+			lane.send_message(REGULAR_PAYLOAD);
+			lane.send_message(REGULAR_PAYLOAD);
+			assert_eq!(lane.storage.data().latest_generated_nonce, 3);
+			assert_eq!(lane.storage.data().latest_received_nonce, 0);
```

Comment:
This is part of every test, I think it would be worthwhile to move it into a function